File: src/MixNetFx/FacessoData/SQLServer/SPAccess_CostCenter.vb
Imports Activedev
Imports Facesso
Imports System.Data
Imports System.Data.SqlClient
Imports System.Data.Common

Partial Public NotInheritable Class SPAccess

    ''' <summary>
    ''' Überprüft, ob die angegebene Kostenstelle für die Subsidiary schon existiert.
    ''' </summary>
    ''' <param name="IDSubsidiary">ID der verwendeten Subsidiary</param>
    ''' <param name="CostCenterNo">Kostenstellennr.</param>
    ''' <returns>True, wenn die Kostenstelle schon vorhanden war.</returns>
    ''' <remarks></remarks>
    Public Function CostCenters_DoesNumberExist(ByVal IDSubsidiary As Integer, _
                                                   ByVal CostCenterNo As Integer, _
                                                   ByVal ExcludeIDCostCenter As ADDBNullable(Of Integer)) _
                                                   As Boolean

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("CostCenters_DoesNumberExist", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = IDSubsidiary
                .Add("@CostCenterNo", SqlDbType.Int).Value = CostCenterNo
                .Add("@ExcludeIDCostCenter", SqlDbType.Int).Value = ExcludeIDCostCenter.Value
                .Add("@DoesExist", SqlDbType.Bit)
                .Item(3).Direction = ParameterDirection.Output
            End With
            locCmd.ExecuteReader()
            Return CBool(locCmd.Parameters(3).Value)
        End Using
    End Function

    ''' <summary>
    ''' Fügt die in CostcenterInfo gespeicherte Kostenstelle der Datenbank hinzu.
    ''' </summary>
    ''' <param name="cci">CostcenterInfo mit den Daten für die Kostenstelle.</param>
    ''' <returns>ErrCode von der Datenbank.</returns>
    ''' <remarks>Die korrelierende Stored Procedure handelt die Versionskontrolle und schließt doppelte Kostenstellennr. aus.</remarks>
    Public Function CostCenters_Add(ByVal cci As CostcenterInfo, ByVal CreatedByIDUser As Integer) As Integer

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("CostCenters_Add", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = cci.IDSubsidiary
                .Add("@CostCenterNo", SqlDbType.Int).Value = cci.CostCenterNo
                .Add("@CostCenterName", SqlDbType.NVarChar, 100).Value = cci.CostCenterName
                .Add("@CostCenterDescription", SqlDbType.NVarChar, 4000).Value = cci.CostCenterDescription.Value
                .Add("@IDCurrency", SqlDbType.NVarChar, 50).Value = cci.IDCurrency
                .Add("@IncentiveIndicatorSynonym", SqlDbType.NVarChar, 50).Value = cci.IncentiveIndicatorSynonym
                .Add("@IncentiveWageSynonym", SqlDbType.NVarChar, 50).Value = cci.IncentiveWageSynonym
                .Add("@IncentiveIndicatorDimension", SqlDbType.NVarChar, 50).Value = cci.IncentiveIndicatorDimension
                .Add("@IncentiveIndicatorPrecision", SqlDbType.TinyInt).Value = cci.IncentiveIndicatorPrecision
                .Add("@UseFixValuedBonus", SqlDbType.Bit).Value = cci.UseFixValuedBonus
                .Add("@IncentiveIndicatorFactor", SqlDbType.Decimal).Value = cci.IncentiveIndicatorFactor
                .Add("@BaseValuePrecision", SqlDbType.TinyInt).Value = cci.BaseValuePrecision
                .Add("@BaseValueSynonym", SqlDbType.NVarChar, 50).Value = cci.BaseValueSynonym
                .Add("@WasCurrentTo", SqlDbType.DateTime).Value = FacessoGeneric.OpenCurrentToDate
                .Add("@CreatedByIDUser", SqlDbType.Int).Value = CreatedByIDUser
                .Add("@IDCostCenterNew", SqlDbType.Int)
                .Item("@IDCostCenterNew").Direction = ParameterDirection.Output
            End With
            locCmd.ExecuteReader()
            Return CInt(locCmd.Parameters("@IDCostCenterNew").Value)
        End Using

    End Function

    Public Function CostCenters_Edit(ByVal cci As CostcenterInfo, ByVal LastEditedByIDUser As Integer) As Integer

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("CostCenters_Edit", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = cci.IDSubsidiary
                .Add("@IDCostCenter", SqlDbType.Int).Value = cci.IDCostCenter
                .Add("@CostCenterNo", SqlDbType.Int).Value = cci.CostCenterNo
                .Add("@CostCenterName", SqlDbType.NVarChar, 100).Value = cci.CostCenterName
                .Add("@CostCenterDescription", SqlDbType.NVarChar, 4000).Value = cci.CostCenterDescription.Value
                .Add("@IDCurrency", SqlDbType.NVarChar, 50).Value = cci.IDCurrency
                .Add("@IncentiveIndicatorSynonym", SqlDbType.NVarChar, 50).Value = cci.IncentiveIndicatorSynonym
                .Add("@IncentiveWageSynonym", SqlDbType.NVarChar, 50).Value = cci.IncentiveWageSynonym
                .Add("@IncentiveIndicatorDimension", SqlDbType.NVarChar, 50).Value = cci.IncentiveIndicatorDimension
                .Add("@IncentiveIndicatorPrecision", SqlDbType.TinyInt).Value = cci.IncentiveIndicatorPrecision
                .Add("@UseFixValuedBonus", SqlDbType.Bit).Value = cci.UseFixValuedBonus
                .Add("@IncentiveIndicatorFactor", SqlDbType.Decimal).Value = cci.IncentiveIndicatorFactor
                .Add("@BaseValuePrecision", SqlDbType.TinyInt).Value = cci.BaseValuePrecision
                .Add("@BaseValueSynonym", SqlDbType.NVarChar, 50).Value = cci.BaseValueSynonym
                .Add("@LastEditedByIDUser", SqlDbType.Int).Value = LastEditedByIDUser
                .Add("@ConsiderHistoryMaintenance", SqlDbType.Bit).Value = ConsiderHistoryMaintenance
                .Add("@IDCostCenterNew", SqlDbType.Int)
                .Item("@IDCostCenterNew").Direction = ParameterDirection.Output
            End With
            locCmd.ExecuteReader()
            Return CInt(locCmd.Parameters("@IDCostCenterNew").Value)
        End Using
    End Function

    Public Function CostCenters_IsInUse(ByVal lvi As CostcenterInfo) As Boolean

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("CostCenters_IsInUse", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDCostCenter", SqlDbType.Int).Value = lvi.IDCostCenter
                .Add("@IDSubsidiary", SqlDbType.Int).Value = lvi.IDSubsidiary
                .Add("@IsInUse", SqlDbType.Bit)
                .Item("@IsInUse").Direction = ParameterDirection.Output
            End With
            locCmd.ExecuteNonQuery()
            Return CBool(locCmd.Parameters("@IsInUse").Value)
        End Using

    End Function

    Public Sub CostCenters_Delete(ByVal lvi As CostcenterInfo)
        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("CostCenters_Delete", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDCostCenter", SqlDbType.Int).Value = lvi.IDCostCenter
                .Add("@IDSubsidiary", SqlDbType.Int).Value = lvi.IDSubsidiary
            End With
            locCmd.ExecuteNonQuery()
        End Using
    End Sub

    Public ReadOnly Property CostCenterInfoItems() As CostcenterInfoItems
        Get
            Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
            If locConnection Is Nothing Then
                Return Nothing
            End If
            Using locConnection
                Dim locCommand As New SqlCommand( _
                "SELECT [CostCenters].*,[Currencies].[CurrencyToken] From [CostCenters] " & _
                "[CostCenters] INNER JOIN [Currencies] ON " & _
                "[CostCenters].[IDCurrency] = [Currencies].[IDCurrency] WHERE " & _
                "[CostCenters].[IDSubsidiary]=" & FacessoGeneric.LoginInfo.IDSubsidiary & _
                " AND [CostCenters].[IsCurrent]='true'", locConnection)
                Dim locDR As SqlDataReader = locCommand.ExecuteReader()
                If locDR.HasRows Then
                    Dim locUIC As New CostcenterInfoItems
                    Do While locDR.Read
                        Dim locCostcenterInfo As New CostcenterInfo(locDR, True)
                        locUIC.Add(locCostcenterInfo)
                    Loop
                    Return locUIC
                Else
                    Return Nothing
                End If
            End Using
        End Get
    End Property

    Public Function GetCostCenter(ByVal IDSubsidiary As Integer, ByVal IDCostCenter As Integer) As CostcenterInfo
        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCommand As New SqlCommand( _
            "SELECT [CostCenters].*,[Currencies].[CurrencyToken] From [CostCenters] " & _
            "[CostCenters] INNER JOIN [Currencies] ON " & _
            "[CostCenters].[IDCurrency] = [Currencies].[IDCurrency] WHERE " & _
            "[CostCenters].[IDSubsidiary]=" & FacessoGeneric.LoginInfo.IDSubsidiary & _
            " AND [CostCenters].[IDCostCenter]=" & IDCostCenter, locConnection)
            Dim locDR As SqlDataReader = locCommand.ExecuteReader()
            If locDR.HasRows Then
                locDR.Read()
                Return New CostcenterInfo(locDR, True)
            Else
                Return Nothing
            End If
        End Using
    End Function

    Public Function GetCurrentBaseCostCenter(ByVal IDSubsidiary As Integer) As CostcenterInfo
        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCommand As New SqlCommand( _
            "SELECT [CostCenters].*,[Currencies].[CurrencyToken] From [CostCenters] " & _
            "[CostCenters] INNER JOIN [Currencies] ON " & _
            "[CostCenters].[IDCurrency] = [Currencies].[IDCurrency] WHERE " & _
            "[CostCenters].[IDSubsidiary]=" & IDSubsidiary & _
            " AND [CostCenters].[IDCostCenterInternal]=" & 0, locConnection)
            Dim locDR As SqlDataReader = locCommand.ExecuteReader()
            locDR.Read()
            Return New CostcenterInfo(locDR, True)
        End Using
    End Function
End Class

File: src/MixNetFx/Facesso.Data/SQLServer/SPAccess_CostCenter.cs
using System;
using System.Data;
using System.Data.SqlClient;
using ActiveDev;
using Facesso;

namespace Facesso.Data
{
    public sealed partial class SPAccess
    {
        public bool CostCenters_DoesNumberExist(int idSubsidiary, int costCenterNo, ADDBNullable<int> excludeIDCostCenter)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return false;
            using (locConnection)
            {
                var locCmd = new SqlCommand("CostCenters_DoesNumberExist", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = idSubsidiary;
                locCmd.Parameters.Add("@CostCenterNo", SqlDbType.Int).Value = costCenterNo;
                locCmd.Parameters.Add("@ExcludeIDCostCenter", SqlDbType.Int).Value = excludeIDCostCenter.Value;
                locCmd.Parameters.Add("@DoesExist", SqlDbType.Bit);
                locCmd.Parameters[3].Direction = ParameterDirection.Output;
                locCmd.ExecuteReader();
                return (bool)locCmd.Parameters[3].Value;
            }
        }

        public int CostCenters_Add(CostcenterInfo cci, int createdByIDUser)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return 0;
            using (locConnection)
            {
                var locCmd = new SqlCommand("CostCenters_Add", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = cci.IDSubsidiary;
                locCmd.Parameters.Add("@CostCenterNo", SqlDbType.Int).Value = cci.CostCenterNo;
                locCmd.Parameters.Add("@CostCenterName", SqlDbType.NVarChar, 100).Value = cci.CostCenterName;
                locCmd.Parameters.Add("@CostCenterDescription", SqlDbType.NVarChar, 4000).Value = cci.CostCenterDescription.Value;
                locCmd.Parameters.Add("@IDCurrency", SqlDbType.NVarChar, 50).Value = cci.IDCurrency;
                locCmd.Parameters.Add("@IncentiveIndicatorSynonym", SqlDbType.NVarChar, 50).Value = cci.IncentiveIndicatorSynonym;
                locCmd.Parameters.Add("@IncentiveWageSynonym", SqlDbType.NVarChar, 50).Value = cci.IncentiveWageSynonym;
                locCmd.Parameters.Add("@IncentiveIndicatorDimension", SqlDbType.NVarChar, 50).Value = cci.IncentiveIndicatorDimension;
                locCmd.Parameters.Add("@IncentiveIndicatorPrecision", SqlDbType.TinyInt).Value = cci.IncentiveIndicatorPrecision;
                locCmd.Parameters.Add("@UseFixValuedBonus", SqlDbType.Bit).Value = cci.UseFixValuedBonus;
                locCmd.Parameters.Add("@IncentiveIndicatorFactor", SqlDbType.Decimal).Value = cci.IncentiveIndicatorFactor;
                locCmd.Parameters.Add("@BaseValuePrecision", SqlDbType.TinyInt).Value = cci.BaseValuePrecision;
                locCmd.Parameters.Add("@BaseValueSynonym", SqlDbType.NVarChar, 50).Value = cci.BaseValueSynonym;
                locCmd.Parameters.Add("@WasCurrentTo", SqlDbType.DateTime).Value = FacessoGeneric.OpenCurrentToDate;
                locCmd.Parameters.Add("@CreatedByIDUser", SqlDbType.Int).Value = createdByIDUser;
                locCmd.Parameters.Add("@IDCostCenterNew", SqlDbType.Int);
                locCmd.Parameters["@IDCostCenterNew"].Direction = ParameterDirection.Output;
                locCmd.ExecuteReader();
                return (int)locCmd.Parameters["@IDCostCenterNew"].Value;
            }
        }

        public int CostCenters_Edit(CostcenterInfo cci, int lastEditedByIDUser)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return 0;
            using (locConnection)
            {
                var locCmd = new SqlCommand("CostCenters_Edit", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = cci.IDSubsidiary;
                locCmd.Parameters.Add("@IDCostCenter", SqlDbType.Int).Value = cci.IDCostCenter;
                locCmd.Parameters.Add("@CostCenterNo", SqlDbType.Int).Value = cci.CostCenterNo;
                locCmd.Parameters.Add("@CostCenterName", SqlDbType.NVarChar, 100).Value = cci.CostCenterName;
                locCmd.Parameters.Add("@CostCenterDescription", SqlDbType.NVarChar, 4000).Value = cci.CostCenterDescription.Value;
                locCmd.Parameters.Add("@IDCurrency", SqlDbType.NVarChar, 50).Value = cci.IDCurrency;
                locCmd.Parameters.Add("@IncentiveIndicatorSynonym", SqlDbType.NVarChar, 50).Value = cci.IncentiveIndicatorSynonym;
                locCmd.Parameters.Add("@IncentiveWageSynonym", SqlDbType.NVarChar, 50).Value = cci.IncentiveWageSynonym;
                locCmd.Parameters.Add("@IncentiveIndicatorDimension", SqlDbType.NVarChar, 50).Value = cci.IncentiveIndicatorDimension;
                locCmd.Parameters.Add("@IncentiveIndicatorPrecision", SqlDbType.TinyInt).Value = cci.IncentiveIndicatorPrecision;
                locCmd.Parameters.Add("@UseFixValuedBonus", SqlDbType.Bit).Value = cci.UseFixValuedBonus;
                locCmd.Parameters.Add("@IncentiveIndicatorFactor", SqlDbType.Decimal).Value = cci.IncentiveIndicatorFactor;
                locCmd.Parameters.Add("@BaseValuePrecision", SqlDbType.TinyInt).Value = cci.BaseValuePrecision;
                locCmd.Parameters.Add("@BaseValueSynonym", SqlDbType.NVarChar, 50).Value = cci.BaseValueSynonym;
                locCmd.Parameters.Add("@LastEditedByIDUser", SqlDbType.Int).Value = lastEditedByIDUser;
                locCmd.Parameters.Add("@ConsiderHistoryMaintenance", SqlDbType.Bit).Value = FacessoGeneric.ConsiderHistoryMaintenance;
                locCmd.Parameters.Add("@IDCostCenterNew", SqlDbType.Int);
                locCmd.Parameters["@IDCostCenterNew"].Direction = ParameterDirection.Output;
                locCmd.ExecuteReader();
                return (int)locCmd.Parameters["@IDCostCenterNew"].Value;
            }
        }

        public bool CostCenters_IsInUse(CostcenterInfo lvi)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return false;
            using (locConnection)
            {
                var locCmd = new SqlCommand("CostCenters_IsInUse", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDCostCenter", SqlDbType.Int).Value = lvi.IDCostCenter;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = lvi.IDSubsidiary;
                locCmd.Parameters.Add("@IsInUse", SqlDbType.Bit);
                locCmd.Parameters["@IsInUse"].Direction = ParameterDirection.Output;
                locCmd.ExecuteNonQuery();
                return (bool)locCmd.Parameters["@IsInUse"].Value;
            }
        }

        public void CostCenters_Delete(CostcenterInfo lvi)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return;
            using (locConnection)
            {
                var locCmd = new SqlCommand("CostCenters_Delete", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDCostCenter", SqlDbType.Int).Value = lvi.IDCostCenter;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = lvi.IDSubsidiary;
                locCmd.ExecuteNonQuery();
            }
        }

        public CostcenterInfoItems CostCenterInfoItems
        {
            get
            {
                SqlConnection locConnection = GetOpenedConnectionSafely();
                if (locConnection == null) return null;
                using (locConnection)
                {
                    var locCommand = new SqlCommand(
                        "SELECT [CostCenters].*,[Currencies].[CurrencyToken] From [CostCenters] " +
                        "[CostCenters] INNER JOIN [Currencies] ON " +
                        "[CostCenters].[IDCurrency] = [Currencies].[IDCurrency] WHERE " +
                        "[CostCenters].[IDSubsidiary]=@IDSubsidiary" +
                        " AND [CostCenters].[IsCurrent]=1", locConnection);
                    locCommand.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = FacessoGeneric.LoginInfo.IDSubsidiary;
                    SqlDataReader locDR = locCommand.ExecuteReader();
                    if (locDR.HasRows)
                    {
                        var locUIC = new CostcenterInfoItems();
                        while (locDR.Read())
                        {
                            var locCostcenterInfo = new CostcenterInfo(locDR, true);
                            locUIC.Add(locCostcenterInfo);
                        }
                        return locUIC;
                    }
                    return null;
                }
            }
        }

        public CostcenterInfo GetCostCenter(int idSubsidiary, int idCostCenter)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return null;
            using (locConnection)
            {
                var locCommand = new SqlCommand(
                    "SELECT [CostCenters].*,[Currencies].[CurrencyToken] From [CostCenters] " +
                    "[CostCenters] INNER JOIN [Currencies] ON " +
                    "[CostCenters].[IDCurrency] = [Currencies].[IDCurrency] WHERE " +
                    "[CostCenters].[IDSubsidiary]=@IDSubsidiary" +
                    " AND [CostCenters].[IDCostCenter]=@IDCostCenter", locConnection);
                locCommand.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = idSubsidiary;
                locCommand.Parameters.Add("@IDCostCenter", SqlDbType.Int).Value = idCostCenter;
                SqlDataReader locDR = locCommand.ExecuteReader();
                if (locDR.HasRows)
                {
                    locDR.Read();
                    return new CostcenterInfo(locDR, true);
                }
                return null;
            }
        }

        public CostcenterInfo GetCurrentBaseCostCenter(int idSubsidiary)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return null;
            using (locConnection)
            {
                var locCommand = new SqlCommand(
                    "SELECT [CostCenters].*,[Currencies].[CurrencyToken] From [CostCenters] " +
                    "[CostCenters] INNER JOIN [Currencies] ON " +
                    "[CostCenters].[IDCurrency] = [Currencies].[IDCurrency] WHERE " +
                    "[CostCenters].[IDSubsidiary]=@IDSubsidiary" +
                    " AND [CostCenters].[IDCostCenterInternal]=@IDCostCenterInternal", locConnection);
                locCommand.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = idSubsidiary;
                locCommand.Parameters.Add("@IDCostCenterInternal", SqlDbType.Int).Value = 0;
                SqlDataReader locDR = locCommand.ExecuteReader();
                locDR.Read();
                return new CostcenterInfo(locDR, true);
            }
        }
    }
}
